A40-csds-support.md

```diff
@@ -0,0 +1,275 @@
+used to query control planes for the synced xDS config of a particular sidecar
+proxy. However, it can also be used to query an xDS-compliant application for
+its received xDS configuration. This doc proposes a solution to implement a CSDS
+service in gRPC, so our users can debug their service mesh easily.
+
+
+## Background
+
+Envoy [started](https://github.com/envoyproxy/envoy/pull/9383) the CSDS
+development process in Dec 2019. The CSDS API has been available since xDS v2,
+and it’s under active development. But, Envoy proxies do not support or
+understand this service, only the control plane does. Envoy already provides
+config dump via its [admin
+interface](https://www.envoyproxy.io/docs/envoy/latest/operations/admin) for
+years. There is no visible plan for Envoy to support serving CSDS directly as a
+proxy (see [envoy#13181](https://github.com/envoyproxy/envoy/issues/13181)).
+
+The CSDS service accepts two methods, one for unary, one for streaming. The
+expected behavior of the two methods is the same, but the streaming method can
+reuse the stream. According to the service description below, this is not a
+watch-style API.
+
+```proto
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// an xDS-compliant client from the management server's point of view. It can
+// also be used to get the current xDS states directly from the client.
+service ClientStatusDiscoveryService {
+  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {
+  }
+
+  rpc FetchClientStatus(ClientStatusRequest) returns (ClientStatusResponse) {
+  }
+}
+```
+
+### Related Proposals: 
+* [A27 - xDS-Based Global Load
+  Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [A38 - Admin Interface API](https://github.com/grpc/proposal/pull/218)
+
+
+## Proposal
+
+### CSDS in gRPC
+
+The goal of adding a CSDS service to gRPC is to enable programmatic access to
+the operating xDS configs of a running gRPC application. To make this more
+clear, an xDS-compliant application receives many xDS configs, which may be
+rejected or ignored or obsoleted. **The gRPC CSDS service should always return
+the currently accepted xDS configs.**
+
+
+### ADS Parsing Logic Update: Continue After First Error
+
+Current gRPC ADS parsing logic is when the response parser observes an error in
+the ADS response, it fails the entire message and aborts parsing.
+
+To improve the debuggability of gRPC, gRPC needs to populate **the reason** and
+**affected resources** when rejecting an update response. This information
+requires the ADS parser to continue parsing past the first error. Here are the
+expected behavior under scenarios:
+
+* If the entire message won’t parse: no need to record anything in CSDS, since
+  we don’t know what type of xDS config or what resources are being updated;
+* If one resource won’t parse: don’t abort the parsing, record the error in the
```

Comment:
The handling is still a bit ambiguous. I guess we would only record the first error, so the parsing does not need to continue. But would attach the error message to all previously accepted resources that the current response is attempting to update.

Reply:
Can you specify the cases that are ambiguous here? I guess the wording here is not accurate enough, if one resource (e.g., a cluster) has validation error, the parsing should continue; or one of its field has parsing error, e.g. a config in Any failed to deserialize, the parsing should continue. In other word, the parsing should continue when possible.

If you need a reference PR, please see https://github.com/grpc/grpc/pull/25329.

Reply:
Looks C-core is [concatenating](https://github.com/markdroth/grpc/blob/0f9af3c045bd9a7d71348e3bd294565386c81c16/src/core/ext/xds/xds_api.cc#L1472) errors together if there are more than one error resources. I don't see much benefit of doing so, other than showing a giant error message while may still not showing all errors (e.g., multiple validation errors for a single resource). Anyway, I am fine with what's being implemented in C-core.

Reply:
I think the important part is peeking into the ADS response to see what set of resources is being updated, then pasting the error details to their `error_state` field. I'm fine with one error, as long as they are logged in the right place in the CSDS response, so it's actionable for users.